### PR TITLE
response to IRASA #1546

### PR DESCRIPTION
**Pull request by ruiliu-psy**
Update20201211: Replace the previous implementation (the copyright remained) with the new one.
Update20201204: As @robertoostenveld suggested, 1) the names of files has been changed and 2) the variable and documentation (including the copyright) has been cleaned up.

New features of this version:
1) This version corrected the [computational order](https://user-images.githubusercontent.com/71086397/93338911-3b46cf80-f82b-11ea-8f44-d8659f6d9f80.jpeg) of the old one (the major concern of  #1546); addressed the[ offset issue at higher frequency](https://user-images.githubusercontent.com/71086397/94351551-1ee03980-005a-11eb-9349-f7441a706609.png).
2) The relevant parts of ft_specest_mtfft were moved into ft_specest_irasa_new, which returns the fractal and original spectra given the cfg.output option.
3) It remained comparable frequency resolution in the fractal and original spectra, by sub-segmenting time series inside ft_specest_irasa and giving equal length of data for estimating the fractal and original spectra.

Further validation needed on the consistency of FieldTrip cfg options. So far it run all right on my local PC.
As for details like documentation, I'd like to first do a internal review of core code. @robertoostenveld @schoffelen may I book the next MEG hackathon?

### CODE REVIEW

**Comment by StolkArjen** [comment]
Thanks for reviving this issue, Rui. A couple of quick questions as I struggle to see what this PR entails:

Per 1, what changes specifically are responsible for correcting for the so-called offset at higher frequencies?
Per 2, is mtmfft still necessary now that irasa outputs both cfg.output = 'powfrac' and 'pow'? Why not have irasa output just the fractal component as before?
Per 3, can you show that the mtmfft and irasa computations are comparable still? Also, is mtmfft producing identical output to irasa with 'pow'?

Additionally, please check your indentation settings (there are seemingly several code change suggestions that seem to be caused by indentation, making it hard to see the changes in the first place), and please don't remove crucial documentation including source citation/explanation and copyrights.

**Comment by ruiliu-psy** [comment]
Per 1, what changes specifically are responsible for correcting the so-called offset at higher frequencies?
See the original issue titled "segmentation and **computational order** in ft_specest_irasa" and responses.  #1546 

> Per 2, is mtmfft still necessary now that irasa outputs both cfg.output = 'powfrac' and 'pow'? Why not have irasa output just the fractal component as before?
Don't have to call ft_specest_mtmfft specifically (see the testing script test/test_issue1546.m).

> Per 3, can you show that the mtmfft and irasa computations are comparable still? 
See #1546 

> Also, is mtmfft producing identical output to irasa with 'pow'? I'm asking because it seems you've moved the sub-segmentation (90% of original segment size) inside irasa, but this would result in disparate frequency resolutions across the irasa and mtmfft methods (the main issue all along)?
Don't need ft_specest_mtmfft now, the whole structure has been changed.
The goal, to me, is not to replicate identical results returned by ft_specest_mtmfft, but to have estimate the fractal and original components with comparable frequency resolution. The new ft_specest_irasa does, given the equal length of sub-segments for 'fracpow' and 'pow' estimations.
Besides frequency resolution and computational order, computing the original component with ft_specest_mtmfft induced another potential biasing between fractal and original components, as irasa is designed to computed auto-power.

> Additionally, please check your indentation settings (there are seemingly several code change suggestions that seem to be caused by indentation, making it hard to see the changes in the first place), and please don't remove crucial documentation including source citation/explanation and copyrights.
will do later

**Comment by StolkArjen** [comment]
> Don't need ft_specest_mtmfft now, the whole structure has been changed. The goal, to me, is not to replicate identical results returned by ft_specest_mtmfft, but to have estimate the fractal and original components with comparable frequency resolution.

The intention of the original implementation was to encapsulate the IRASA method within fieldtrip such that it is compatible with the other spectral estimation functions. For instance, by virtue of this PR, users would get inconsistent results across the mtmfft and irasa methods, and there's no documentation to suggest otherwise or help to understand those differences. If Robert and/or JM agreed, I wouldn't mind deviating from this intention for the purpose of parity with Wen & Liu as long as it's super-clear that the irasa output is no longer interchangeable with other output. At the same time, it's not entirely obvious to me 1) _what's_ driving this "offset in higher frequencies", and 2) _whether_ it's worth introducing the undocumented inconsistency suggested by this PR.

**Comment by ruiliu-psy** [comment]
I don't understand what you meant by 'inconsistent results across the mtmfft and irasa methods' and 'inconsistency'. If you are talking about output structure, they are consistent with other output of ft_specest functions. The new implementation included the relevant parts of ft_specest_mtmfft (except for phase correction) for computing the original component. If one wants to subtract fractal from original, I think it would be more appropriate to subtract two auto-power spectrums (This was also how it's been done in Wen&Liu, 2016), rather than one auto- vs. cross-spectrum. The estimation of original component by the new implementation are very close to that by ft_specest_mtmfft as I've replied in the last post (see # [1546](https://github.com/fieldtrip/fieldtrip/issues/1546#issuecomment-734926120) red and blue solid line were almost on top of each other)

As for the documentation, I'm sorry! Robert deleted everything after line13. I think he just intended to show me the general framework of ft functions. This PR is made for the MEG hackathon review, we will put everything documentation back after the testing.

**Comment by robertoostenveld** [comment]
I just made some improvements to the code. 

Looking at the code, I wonder why you call it `mixed/fractal` in the algorithm and in the `cfg.method`? I understand fractal, but mixed is a feature of the signal that you put into it, not a feature of the algorithm. I propose to call it `original/fractal`, indicating that it returns the original power spectrum (using the sub-segmenting that is inherent to IRASA) or the fractal component (with the same sub-segmenting).

**Comment by ruiliu-psy** [comment]
I've made a change to the files.
The naming was suggested by Arjen for consistency of Wen and Liu's. Either mixed/fractal or original/fractal is more intuitive than fracpow/pow, I think. Thank you for guiding and reviewing this work!

File: ft_freqanalysis.m [hunk 1]
```diff
@@ -288,13 +288,13 @@
...
     cfg.taper       = ft_getopt(cfg, 'taper', 'hanning');
-    cfg.output      = ft_getopt(cfg, 'output', 'pow');
+    cfg.output      = ft_getopt(cfg, 'output', 'fractal');
     cfg.pad         = ft_getopt(cfg, 'pad', 'nextpow2');
     if ~isequal(cfg.taper, 'hanning')
       ft_error('the irasa method supports hanning tapers only');
     end
-    if ~isequal(cfg.output, 'pow')
-      ft_error('the irasa method outputs power only');
+    if ~strcmp(cfg.output, {'fractal','original'})
```
Comment (robertoostenveld) [inline ft_freqanalysis.m:296]: I suggest to move this check and error message to the `ft_spectest_irasa` function, which gets the desired `output` as  input argument.

**Review by robertoostenveld — COMMENTED** [review]
I made some edits and have one small suggestion regarding the error handling. Other than that it looks ok and could be merged.

**Comment by robertoostenveld** [comment]
@StolkArjen perhaps you wanna have a look at this?

File: ft_freqanalysis.m (hunk 1 above)
Comment (ruiliu-psy) [reply ft_freqanalysis.m]: done!

**Comment by ruiliu-psy** [comment]
> @StolkArjen perhaps you wanna have a look at this?

Hi, Robert! I invited Arjen for a review and he had agreed with the changes and replacement as I told you on mattermost.

**Comment by StolkArjen** [comment]
Hi both, I won't be able to actually test it anytime soon, but the code and doc borh look clean to me, and I've reviewed past versions with Rui. I'd appreciate it though if prior efforts were acknowledged (in the copyrights) given that the function directly builds on and replaces the old implementation. As for the example wiki page, it's hard to see what it would look like when finished. It would be great to demonstrate the use of IRASA on the sensorimotor-ECoG dataset (the example page originates from the supplemental material of that study), but I could finish / add in at a later stage myself (hopefully).

**Pull request merged**